content/posts/2022-09-14-introducing-jaiminho/index.md

```diff
@@ -0,0 +1,136 @@
+---
+title: Introducing Jaiminho, an implementation of the Transactional Outbox Pattern for Django
+author: Rafael Zanetti
+twitter: LoadsmartEng
+layout: post
+lang: en
+path: /blog/introducing-jaiminho
+date: 2022-09-14
+comments: true
+---
+
+At Loadsmart, many of our services are organized within an event-driven architecture and face some usual challenges that come with such an architecture. One of them is **dual writes**, which happens when an application has to change data on two different places. A common example of dual writes is an application that needs to store data on a database and notify that change by emitting an event which may be consumed by different services. The inherent issue with this scenario is that if one of these two operations fails, the producer and consumer applications will have data inconsistencies, because the consumer never received the event informing the database change.
+
+There are well-known methods to deal with such issues, and we chose to apply the transactional outbox pattern. Since we couldn’t find any related solution on [Django community](https://djangopackages.org/search/?q=outbox), we decided to implement an open-source library called [Jaiminho](https://github.com/loadsmart/jaiminho), which is a broker agnostic implementation of the outbox pattern.
+
+
+# The Transactional outbox pattern
+
+For simplicity purposes, let’s assume the common case of an application that has to save data to a database and notify it to an event stream. The idea behind the transactional outbox pattern is combining both writes into a single atomic transaction, which guarantees that if one of them fails, the entire transaction fails, avoiding inconsistent scenarios; in this case, it means that if a record is persisted into the database, it is guaranteed to be notified to the stream. The pattern achieves this goal by introducing two concepts:
+
+- Outbox table
+
+When storing a record to the database, the corresponding event should be persisted into an outbox table, and both inserts should be in the same transaction. With this strategy we can leverage the atomicity property of relational databases transactions to ensure either both or no inserts will be successful.
+
+- Message relay
+
+The outbox table is guaranteed to have all events that should be notified, so we just have to ensure that they are really sent. This responsibility goes to the message relay, which will publish all events in the table to the stream and mark them as sent.
+Note: Since the message relay has to send an event and then mark it as sent, that characterizes another dual write, because if the relay fails to update the status after sending the event, it will be sent again on the next execution. So, to ensure no duplicated events are delivered, the message consumers must implement idempotence on their side.
```

Comment:
`So, to ensure no duplicated events are delivered` maybe we could change delivered to consumed here

Reply:
Done!